### DISCUSSION TITLE
Fabric turtle/pocket upgrades registration in 1.19.4

### DISCUSSION THREAD

**Discussion by SirEdvin** [cc-tweaked/CC-Tweaked#1432] [Q&A] [answered]
For forge turtle upgrades registration is fairly straightforward. However, when it comes to fabric, how do I suppose to register turtle upgrades?

I, obviously, need access to created registries, because created my own will not help (fabric in general mostly about registry extending that creating it own), but for now [code](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.19.x/projects/fabric/src/main/java/dan200/computercraft/shared/ComputerCraft.java#L53) doesn't store them anywhere and I am forced to retrieve them dynamically.

And very strange thing about this, that this code:

```kotlin
println(RegistriesAccessor.getROOT().keySet())
val turtleSerializerRegister = RegistriesAccessor.getROOT().get(TurtleUpgradeSerialiser.REGISTRY_ID.location()) as Registry<TurtleUpgradeSerialiser<*>>
val pocketSerializerRegister = RegistriesAccessor.getROOT().get(PocketUpgradeSerialiser.REGISTRY_ID.location()) as Registry<PocketUpgradeSerialiser<*>>
```

works, but this

```kotlin
val turtleSerializerRegister = RegistriesAccessor.getROOT().get(TurtleUpgradeSerialiser.REGISTRY_ID.location()) as Registry<TurtleUpgradeSerialiser<*>>
val pocketSerializerRegister = RegistriesAccessor.getROOT().get(PocketUpgradeSerialiser.REGISTRY_ID.location()) as Registry<PocketUpgradeSerialiser<*>>
```

Throws NullPointerException, because `RegistriesAccessor.getROOT().get(TurtleUpgradeSerialiser.REGISTRY_ID.location())` returns null instead of proper register (even if computercraft init was called first).

PS: And of course, this behavior inconsistent!

**Comment by SirEdvin** [comment, marked as answer]
Okey, I have answer now. You should use RegistryEntryAddedCallback event to resolve this issue.

So, when you trying to access registries and they are not available, you should register callback for RegistryEntryAddedCallback, **but** if registries are available, you must register with them, because you will not have any RegistryEntryAddedCallback event for them.

**Reply by SquidDev** [reply]
Not answering your question, but can you not use `BuiltInRegistries.REGISTRY.get(TurtleUpgradeSerialiser.REGISTRY_ID.location());` instead of ` RegistriesAccessor.getROOT()` (which I'm assuming is an accessor mixin)? That's what I do here:

https://github.com/cc-tweaked/CC-Tweaked/blob/3bf29695fb2544e74892672ce17b7d7444487a8f/projects/fabric/src/main/java/dan200/computercraft/shared/platform/PlatformHelperImpl.java#L90-L95

**Reply by SirEdvin** [reply]
Yes, I also should do this, thanks!

**Comment by SirEdvin** [comment]
Okey, code still goes out of control with this problem

```kotlin
PeripheralWorksPlatform.registerTurtleUpgrade(
            PeripheraliumHubPeripheral.ID,

            TurtleUpgradeSerialiser.simpleWithCustomItem {_, stack ->
                return@simpleWithCustomItem PeripheraliumHubTurtleUpgrade(PeripheralWorksConfig.peripheraliumHubUpgradeCount, stack)
            }, { dataProvider, serializer ->
                dataProvider.simpleWithCustomItem(
                    PeripheraliumHubPeripheral.ID,
                    serializer, Items.PERIPHERALIUM_HUB.get()
                )
            },
            listOf(Consumer{
                PeripheralWorksClientCore.registerHook { ComputerCraftAPIClient.registerTurtleUpgradeModeller(
                    it.get(), TurtleUpgradeModeller.flatItem()
                ) }
            })
        )
```

@SquidDev can we have some special API for fabric for registration update in registries or just this registries exposed directly?
